### PR TITLE
Tighten launcher and CUDA compat guards

### PR DESCRIPTION
**Pull request by 1larity**
## Summary
- treat unexpected CUDA bool-argsort probe failures as unsupported
- stop the Windows launchers immediately when the legacy torch compat step fails
- propagate unexpected legacy-compat probe failures from the Linux API launcher

## Stacking note
- this branch is stacked on top of #827 today
- until #827 is merged, GitHub will show the combined diff against main rather than only this top commit

## Validation
- uv run python -m unittest acestep.core.generation.handler.init_service_test acestep.launcher_legacy_torch_fix_test

## Summary by CodeRabbit

* **New Features**
  * CLI accepts normalized quantization values (int8_weight_only, fp8_weight_only, w8a8_dynamic, none) with an auto-detected default.
  * Startup scripts include a probe-and-fix flow to detect and optionally install legacy NVIDIA PyTorch support.

* **Improvements**
  * Safer, tolerant CUDA and quantization probing to avoid startup failures.
  * Quantization no longer forces model compilation.
  * Improved bfloat16 detection and device-aware dtype selection.

* **Tests**
  * New unit tests for quantization parsing, CUDA/device probing, compatibility probe, and startup guards.

### CODE REVIEW

File: acestep/acestep_v15_pipeline.py [hunk 1]
```diff
@@ -305,17 +307,26 @@ def main():
...
+    if gpu_config.quantization_default and not _is_mac:
+        _default_quantization = "int8_weight_only"
+        try:
+            import torch
+            if torch.cuda.is_available():
+                major, _ = torch.cuda.get_device_capability(0)
+                if major < 7:
+                    _default_quantization = "w8a8_dynamic"
+        except Exception:
+            pass
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/acestep_v15_pipeline.py]: Handled in `b68cac0`. The CUDA capability probe now logs a warning on probe failure instead of swallowing the exception, while preserving the quantization fallback behavior.

File: acestep/core/generation/handler/init_service_loader.py [hunk 2]
```diff
@@ -10,6 +11,101 @@
...
+            target_device = "cuda"
+        try:
+            _ = torch.tensor([True, False], device=target_device).argsort()
+            return True
+        except RuntimeError as exc:
+            logger.debug(
+                "[_cuda_supports_bool_argsort] Treating CUDA bool argsort probe failure as unsupported: {}",
+                exc,
+            )
+            return False
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/init_service_loader.py:30]: Handled in `b68cac0`. `_cuda_supports_bool_argsort()` now probes the same 2-D `argsort(dim=1, descending=True, stable=True)` path that `pack_sequences()` uses, and the regression test asserts that exact call.

File: acestep/core/generation/handler/init_service_loader.py [hunk 3]
```diff
@@ -10,6 +11,101 @@
...
+        quant_config = self._build_quantization_config(quantization)
+
+        def _dit_filter_fn(module, fqn):
+            """Keep only DiT linear layers and exclude tokenizer/detokenizer paths."""
+            if not _is_linear(module, fqn):
+                return False
+            for part in fqn.split("."):
+                if part in ("tokenizer", "detokenizer"):
+                    return False
+            return True
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/init_service_loader.py:107]: Handled in `b68cac0`. `_dit_filter_fn()` now restricts quantization to `decoder.*` and still excludes tokenizer/detokenizer paths. I also added a focused test for the filter behavior.

File: acestep/core/generation/handler/init_service_memory_basic.py [hunk 4]
```diff
@@ -66,12 +66,12 @@ def _get_affine_quantized_tensor_class():
...
             return AffineQuantizedTensor
-        except ImportError:
+        except Exception:
             pass
         try:
             from torchao.quantization.affine_quantized import AffineQuantizedTensor
             return AffineQuantizedTensor
-        except ImportError:
+        except Exception:
             pass
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/init_service_memory_basic.py]: Handled in `b68cac0`. Both `AffineQuantizedTensor` import probes now log debug details instead of silently swallowing compatibility failures.

File: acestep/core/generation/handler/memory_utils.py [hunk 5]
```diff
@@ -6,7 +6,22 @@
...
+
+
+def _is_cuda_device(device: object) -> bool:
+    """Return whether a device identifier refers to any CUDA device."""
+    if device is None:
+        return False
+    try:
+        return torch.device(str(device)).type == "cuda"
+    except (TypeError, RuntimeError, ValueError):
+        return isinstance(device, str) and device.split(":", 1)[0] == "cuda"
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/memory_utils.py:25]: Handled in `b68cac0`. The VRAM helpers now extract and forward the selected CUDA device index, and there are focused tests covering `cuda:1`.

File: acestep/launcher_compat.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    """Return launcher probe exit code for legacy torch compatibility check.
+
+    Args:
+        torch_module: Optional injected ``torch`` module for tests.
+
+    Returns:
+        ``LEGACY_TORCH_FIX_EXIT_CODE`` when fix should be applied, otherwise ``0``.
+    """
+    decision = determine_legacy_torch_fix(torch_module=torch_module)
+    return LEGACY_TORCH_FIX_EXIT_CODE if decision.should_apply else 0
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/launcher_compat.py]: Handled in `b68cac0`. `legacy_torch_fix_probe_exit_code()` now treats `probe_failed` as a non-zero fail-fast result.

File: acestep/launcher_legacy_torch_fix_test.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+    def test_windows_api_launcher_calls_shared_probe(self) -> None:
+        """Windows API launcher should call shared Python compatibility probe."""
+        content = self._read("start_api_server.bat")
+        self.assertIn('if /i "%ACESTEP_SKIP_LEGACY_TORCH_FIX%"=="true"', content)
+        self.assertIn("legacy_torch_fix_probe_exit_code", content)
+        self.assertIn("torch==2.5.1+cu121", content)
+        self.assertEqual(content.count("call :EnsureLegacyNvidiaTorchCompat"), 1)
+        self.assertEqual(content.count("if !ERRORLEVEL! NEQ 0 exit /b !ERRORLEVEL!"), 1)
+
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/launcher_legacy_torch_fix_test.py:68]: Handled in `b68cac0`. The launcher-script tests now use presence checks instead of exact counts and assert the probe call is immediately followed by exit-code propagation.

File: acestep/ui/gradio/events/generation/service_init_test.py [hunk 8]
```diff
@@ -250,5 +251,50 @@ def test_init_llm_with_auto_device_calls_initialize(self, mock_gpu_config):
...
+            "acestep.ui.gradio.events.generation.service_init"
+        )
+        real_import = __import__
+
+        def fake_import(name, globals=None, locals=None, fromlist=(), level=0):
+            """Raise ImportError only for torch imports from the helper."""
+            if name == "torch":
+                raise ImportError("torch missing")
+            return real_import(name, globals, locals, fromlist, level)
+
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/ui/gradio/events/generation/service_init_test.py]: Handled in `b68cac0`. The mocked import helper now uses `globals_`/`locals_` so it no longer shadows builtins.

File: start_api_server.bat [hunk 9]
```diff
@@ -270,6 +270,9 @@ if exist "%~dp0python_embeded\python.exe" (
         echo.
     )
 
+    call :EnsureLegacyNvidiaTorchCompat
+    if !ERRORLEVEL! NEQ 0 exit /b !ERRORLEVEL!
```
[the remark this answers is not in the document]
Comment (1larity) [reply start_api_server.bat:274]: Handled in `2059184`. `start_api_server.bat` now uses `uv run --no-sync` for both the normal and offline retry paths so the legacy torch override is not undone before launch.

File: start_api_server.bat [hunk 10]
```diff
@@ -305,3 +308,41 @@ if exist "%~dp0python_embeded\python.exe" (
...
+    ) else (
+        echo [Compatibility] Warning: failed to install torchao==0.11.0. Quantization may be unavailable.
+    )
+) else (
+    echo [Compatibility] Warning: automatic legacy torch install failed.
+    echo [Compatibility] Run manually:
+    echo   uv pip install --python .venv\Scripts\python.exe --force-reinstall --index-url https://download.pytorch.org/whl/cu121 torch==2.5.1+cu121 torchvision==0.20.1+cu121 torchaudio==2.5.1+cu121
+)
+popd
+exit /b 0
```
[the remark this answers is not in the document]
Comment (1larity) [reply start_api_server.bat]: Handled in `2059184`. `:EnsureLegacyNvidiaTorchCompat` in `start_api_server.bat` now propagates non-zero probe and install failures instead of falling through as success.

File: start_gradio_ui.bat [hunk 11]
```diff
@@ -344,6 +347,41 @@ goto :eof
...
+    ) else (
+        echo [Compatibility] Warning: failed to install torchao==0.11.0. Quantization may be unavailable.
+    )
+) else (
+    echo [Compatibility] Warning: automatic legacy torch install failed.
+    echo [Compatibility] Run manually:
+    echo   uv pip install --python .venv\Scripts\python.exe --force-reinstall --index-url https://download.pytorch.org/whl/cu121 torch==2.5.1+cu121 torchvision==0.20.1+cu121 torchaudio==2.5.1+cu121
+)
+popd
+exit /b 0
```
[the remark this answers is not in the document]
Comment (1larity) [reply start_gradio_ui.bat]: Handled in `b68cac0`. The Gradio batch launcher now propagates probe/install failures out of `:EnsureLegacyNvidiaTorchCompat` instead of returning success.

File: start_gradio_ui.sh [hunk 12]
```diff
@@ -275,6 +275,43 @@ fi
...
+        if (cd "$SCRIPT_DIR" && uv pip install --python .venv/bin/python --force-reinstall torchao==0.11.0 >/dev/null 2>&1); then
+            echo "[Compatibility] Installed torchao==0.11.0 (legacy-compatible)."
+        else
+            echo "[Compatibility] Warning: failed to install torchao==0.11.0. Quantization may be unavailable."
+        fi
+    else
+        echo "[Compatibility] Warning: failed to install legacy torch automatically."
+        echo "[Compatibility] Run manually:"
+        echo "  uv pip install --python .venv/bin/python --force-reinstall --index-url https://download.pytorch.org/whl/cu121 torch==2.5.1+cu121 torchvision==0.20.1+cu121 torchaudio==2.5.1+cu121"
+    fi
```
[the remark this answers is not in the document]
Comment (1larity) [reply start_gradio_ui.sh:320]: Handled in `b68cac0`. The shell launcher now fails closed on probe errors and on failed legacy torch installation, instead of continuing into startup with an incompatible wheel set.

File: acestep/core/generation/handler/init_service_test.py [hunk 13]
```diff
@@ -255,10 +257,18 @@ def test_resolve_initialize_device_requested_cuda_falls_back_to_mps(self):
...
+        """It allows quantization checks even when compile mode is disabled."""
         host = _Host(project_root="K:/fake_root", device="cpu")
-        with self.assertRaises(ValueError):
+        real_import = builtins.__import__
+
+        def fake_import(name, globals=None, locals=None, fromlist=(), level=0):
+            """Return a torchao stub so the check can pass without real dependency."""
+            if name == "torchao":
+                return types.ModuleType("torchao")
+            return real_import(name, globals, locals, fromlist, level)
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/init_service_test.py]: Handled in `b68cac0`. I renamed all four `fake_import` helpers in `init_service_test.py` to use `globals_`/`locals_` and updated their delegate calls accordingly.

File: acestep/launcher_compat_test.py [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+        decision = determine_legacy_torch_fix(types.SimpleNamespace(cuda=broken_cuda))
+        self.assertFalse(decision.should_apply)
+        self.assertEqual("probe_failed", decision.reason)
+
+    def test_probe_exit_code_maps_to_decision(self) -> None:
+        """Probe exit code should be 42 only when legacy fix is required."""
+        needs_fix = _make_torch(cuda_available=True, capability=(6, 1), arch_list=["sm_70"])
+        ok = _make_torch(cuda_available=True, capability=(8, 0), arch_list=["sm_80"])
+        self.assertEqual(LEGACY_TORCH_FIX_EXIT_CODE, legacy_torch_fix_probe_exit_code(needs_fix))
+        self.assertEqual(0, legacy_torch_fix_probe_exit_code(ok))
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/launcher_compat_test.py:87]: Handled in `b68cac0`. The test now covers the `probe_failed` branch as well. In the current contract that branch maps to `LEGACY_TORCH_FIX_EXIT_CODE`, so the assertion checks for a non-zero fail-fast result and that concrete exit code.

File: acestep/core/generation/handler/memory_utils.py [hunk 15]
```diff
@@ -123,12 +155,12 @@ def _vram_guard_reduce_batch(
                 if batch_size <= tier_max:
                     logger.debug(
                         f"[VRAM guard] offload_to_cpu=True, batch_size={batch_size} <= "
-                        f"tier limit {tier_max} — skipping dynamic VRAM check"
+                        f"tier limit {tier_max} Ã¢â‚¬â€ skipping dynamic VRAM check"
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/memory_utils.py]: Handled in `c60e35a`. I replaced the mojibake in that debug message with a plain ASCII hyphen so the log stays readable on Windows and avoids ambiguous-unicode noise.

File: acestep/core/generation/handler/memory_utils.py [hunk 16]
```diff
@@ -142,7 +174,7 @@ def _vram_guard_reduce_batch(
         safety_margin_gb = 1.5
         available_for_batch = free_gb - safety_margin_gb
         if available_for_batch <= 0:
-            logger.warning(f"[VRAM guard] Only {free_gb:.1f} GB free — reducing batch_size to 1")
+            logger.warning(f"[VRAM guard] Only {free_gb:.1f} GB free Ã¢â‚¬â€ reducing batch_size to 1")
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/memory_utils.py]: Handled in `c60e35a`. The warning string now uses a plain ASCII hyphen as well, which avoids the Windows encoding damage and keeps the log output consistent.

File: acestep/launcher_compat_test.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        broken_cuda = types.SimpleNamespace(
+            is_available=lambda: True,
+            get_device_capability=lambda _idx=0: (_ for _ in ()).throw(RuntimeError("boom")),
+            get_arch_list=lambda: [],
+        )
+        self.assertEqual(LEGACY_TORCH_FIX_EXIT_CODE, legacy_torch_fix_probe_exit_code(needs_fix))
+        self.assertEqual(0, legacy_torch_fix_probe_exit_code(ok))
+        failed_exit = legacy_torch_fix_probe_exit_code(types.SimpleNamespace(cuda=broken_cuda))
+        self.assertNotEqual(0, failed_exit)
+        self.assertEqual(LEGACY_TORCH_FIX_EXIT_CODE, failed_exit)
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/launcher_compat_test.py:90]: Handled in `c60e35a`. The docstring now matches the asserted contract: `legacy_torch_fix_probe_exit_code()` returns `LEGACY_TORCH_FIX_EXIT_CODE` for both the required-fix path and the probe-failed path, and `0` for the compatible case.

File: acestep/core/generation/handler/memory_utils.py [hunk 18]
```diff
@@ -157,11 +191,15 @@ def _vram_guard_reduce_batch(
...
         target_device = device or self.device
-        if target_device in ["cuda", "xpu"]:
-            if target_device == "cuda" and is_rocm_available():
+        if _is_cuda_device(target_device):
+            if is_rocm_available():
                 # On ROCm, defer to self.dtype which is already set to a safe
                 # value (float32 by default, or ACESTEP_ROCM_DTYPE override).
                 return getattr(self, "dtype", torch.float32)
+            if cuda_supports_bfloat16():
+                return torch.bfloat16
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/core/generation/handler/memory_utils.py:200]: Handled in `1b7c695`. `cuda_supports_bfloat16()` now accepts an optional device index, `_get_vae_dtype()` passes `_cuda_device_index(target_device)`, and the VAE dtype test now asserts the `cuda:1` path calls the BF16 probe with index `1`.

File: start_gradio_ui.bat [hunk 19]
```diff
@@ -344,6 +347,54 @@ goto :eof
...
+    set "LEGACY_INSTALL_EXIT=!ERRORLEVEL!"
+    echo [Compatibility] Warning: automatic legacy torch install failed.
+    echo [Compatibility] Run manually:
+    echo   uv pip install --python .venv\Scripts\python.exe --force-reinstall --index-url https://download.pytorch.org/whl/cu121 torch==2.5.1+cu121 torchvision==0.20.1+cu121 torchaudio==2.5.1+cu121
+    set "LEGACY_HELPER_EXIT=!LEGACY_INSTALL_EXIT!"
+    popd
+    exit /b !LEGACY_HELPER_EXIT!
+)
+popd
+    exit /b !LEGACY_HELPER_EXIT!
```
[the remark this answers is not in the document]
Comment (1larity) [reply start_gradio_ui.bat]: Handled in `1b7c695`. The Gradio batch helper now exits explicitly with `0` on the success path instead of relying on the unset `LEGACY_HELPER_EXIT` variable.
